src/material_library.cpp

```diff
@@ -0,0 +1,221 @@
+#include <unistd.h>
+#include <iostream>
+
+#ifndef PYNE_IS_AMALGAMATED
+#include "material_library.h"
+#endif
+
+namespace pyne {
+// Empty Constructor
+MaterialLibrary::MaterialLibrary(){};
+
+// Default constructor
+MaterialLibrary::MaterialLibrary(std::string file, std::string datapath) {
+  if (!check_file_exists(file)) {
+    std::cerr << "The file " << file << " does not exist or is read protected"
+              << std::endl;
+    exit(1);
+  }
+  if (!hdf5_path_exists(file, datapath)) {
+    std::cerr << "The datapath " << datapath << " in " << file << " empty."
+              << std::endl;
+    exit(1);
+  }
+
+  // load materials
+  from_hdf5(file);
+};
+
+// Destructor
+MaterialLibrary::~MaterialLibrary(){};
+
+// convert convert a filename into path+filename (for pyne)
+std::string MaterialLibrary::get_full_filepath(char* filename) {
+  std::string file(filename);
+  return MaterialLibrary::get_full_filepath(file);
+}
+
+// convert convert a filename into path+filename (for pyne)
+std::string MaterialLibrary::get_full_filepath(std::string filename) {
+  // remove all extra whitespace
+  filename.erase(std::remove(filename.begin(), filename.end(), ' '),
+                 filename.end());
+  // use stdlib call
+  const char* full_filepath = realpath(filename.c_str(), NULL);
+  return std::string(full_filepath);
+}
+
+// see if file exists
+bool MaterialLibrary::check_file_exists(std::string filename) {
+  // from http://stackoverflow.com/questions/12774207/
+  // fastest-way-to-check-if-a-file-exist-using-standard-c-c11-c
+  std::ifstream infile(filename.c_str());
+  return infile.good();
+}
+
+// Append Material to the library from hdf5 file
+void MaterialLibrary::from_hdf5(std::string filename, std::string datapath,
+                                int protocol) {
+  const char* data_path = datapath.c_str();
+
+  if (!hdf5_path_exists(filename, data_path)) return;
+
+  int num_materials = get_length_of_table(filename, datapath);
+
+  for (int i = 0; i < num_materials; i++) {
+    pyne::Material mat;  // from file
+    // read new mat
+    mat.from_hdf5(filename, datapath, i, protocol);
+
+    append_to_nuclist(mat);
+    material_library[mat.metadata["name"].asString()] = mat;
+    matlist.insert(mat.metadata["name"].asString());
+  }
+}
+
+void MaterialLibrary::add_material(pyne::Material mat) {
+  std::string mat_name = mat.metadata["name"].asString();
+  std::pair<pyne::mat_iter, bool> ins = material_library.insert(
+      std::pair<std::string, pyne::Material>(mat_name, mat));
+
+  if (ins.second) {
+    append_to_nuclist(mat);
+    matlist.insert(mat.metadata["name"].asString());
+  }
+}
+
+void MaterialLibrary::del_material(pyne::Material mat) {
+  std::string mat_name = mat.metadata["name"].asString();
```

Comment:
I honestly can't recall exactly, but I don't think the current MaterialLibrary relies on always having a `name` attribute in it's metadata like this.

https://github.com/pyne/pyne/blob/develop/pyne/material.pyx#L2354

Reply:
you are right! I'll also need to fix the `from_df5` method as well, to generate unique names if required!

Reply:
does it look better now ?

Reply:
no but it generates them when adding the materials into the library, so I believe we have to relie on them to remove material from the library...

Reply:
I think that this version of the PyNE Material library relies on that. In other words, UWUW relies on that, but IIRC PyNE materials are not expected to have names in their metadata.

They aren't even required to have valid densities I don't think :)

Reply:
Ok, I now understand why you are doing such comment... (or at least I believe so...)
Does it look better now ?